File: 2023/13/mirrors.rb
#!/usr/bin/env ruby

def find_reflection(a, direction)
  for rowidx in 0...a.size - 1
    if a[rowidx] == a[rowidx + 1]
      mirrortop = rowidx
      mirrorsize = 1
      while mirrortop > 0 && a[mirrortop - 1] == a[rowidx + mirrorsize + 1]
        mirrortop -= 1
        mirrorsize += 1
      end

      # It only counts if it goes up to an edge.
      if mirrortop == 0 || rowidx + mirrorsize + 1 == a.size
        puts "found #{direction} reflection: #{rowidx} - #{rowidx + 1} of size #{mirrorsize}"
        val = rowidx + 1
        val *= 100 if direction == "horizontal"
        puts "value: #{val}"
        return val
      end
    end
  end
  0
end

def row_edit_distance(a, b)
  fail if a.size != b.size

  edit_distance = 0
  for i in 0...a.size
    if a[i] != b[i]
      edit_distance += 1
    end
  end

  edit_distance
end

def find_horizontal_reflection(a)
  find_reflection(a, "horizontal")
end

def find_smudged_reflection(a)
  for rowidx in 0...a.size - 1
    mirrorsize = 0
    mirror_edit_distance = 0
    x = 0
    while rowidx - mirrorsize >= 0 && rowidx + mirrorsize + 1 < a.size
      x = row_edit_distance(a[rowidx - mirrorsize], a[rowidx + mirrorsize + 1])
      break if mirror_edit_distance + x > 1
      mirror_edit_distance += x
      mirrorsize += 1
    end

    next if mirrorsize == 0
    next if mirror_edit_distance != 1
    # It only counts if it goes up to an edge.
    if rowidx - mirrorsize == -1 || rowidx + mirrorsize + 1 == a.size
      #p "found smudged reflection at #{rowidx} of size #{mirrorsize}"
      return rowidx
    end
  end
  nil
end

def find_smudged_horizontal_reflection(a)
  x = find_smudged_reflection(a)
  if x
    puts "found smudged horizontal reflection, value: #{100 * (x + 1)}"
    return 100 * (x + 1)
  end
  0
end

def find_vertical_reflection(a)
  # rotate the whole array
  cols = a[0].zip(*a[1..])

  find_reflection(cols, "vertical")
end

def find_smudged_vertical_reflection(a)
  # rotate the whole array
  cols = a[0].zip(*a[1..])

  x = find_smudged_reflection(cols)
  if x
    puts "found smudged vertical reflection, value: #{x + 1}"
    return x + 1
  end
  0
end

def get_pattern(f)
  a = []
  while(!f.eof? && line = f.readline)
    break if line == "\n"
    a.push(line.chomp.chars)
  end

  a
end

def main
  totalvalue = 0
  f = open("input.txt")
  aidx = 0
  while(a = get_pattern(f); !a.empty?)
    puts "pattern #{aidx}"
    #totalvalue += find_horizontal_reflection(a)
    #totalvalue += find_vertical_reflection(a)
    totalvalue += find_smudged_horizontal_reflection(a)
    totalvalue += find_smudged_vertical_reflection(a)

    aidx += 1
  end

  puts "total value: #{totalvalue}"
end

main()

File: 2023/13/mirrors.rs
type Pattern = Vec<Vec<char>>;

fn find_reflection(a: &Pattern, direction: &str) -> u64 {
    use std::convert::TryFrom;

    for rowidx in 0..a.len() - 1 {
        if a[rowidx] == a[rowidx + 1] {
            let mut mirrortop = rowidx;
            let mut mirrorsize = 1;
            while mirrortop > 0 && rowidx + mirrorsize + 1 < a.len() && a[mirrortop - 1] == a[rowidx + mirrorsize + 1] {
                mirrortop -= 1;
                mirrorsize += 1;
            }

            // It only counts if it goes up to an edge.
            if mirrortop == 0 || rowidx + mirrorsize + 1 == a.len() {
                println!("found {} reflection: {} - {} of size {}", direction, rowidx, rowidx + 1, mirrorsize);
                let mut val: u64 = u64::try_from(rowidx).unwrap() + 1;
                if direction == "horizontal" {
                    val *= 100;
                }
                println!("value: {}", val);
                return val;
            }
        }
    }
    return 0;
}

fn find_horizontal_reflection(a: &Pattern) -> u64 {
    return find_reflection(&a, "horizontal");
}

fn rotate_pattern(a: &Pattern) -> Pattern {
    let mut cols: Pattern = vec![];
    for i in 0..a[0].len() {
        let mut col = vec![];
        for row in a.iter() {
            col.push(row[i]);
        }
        cols.push(col);
    }
    return cols;
}

fn find_vertical_reflection(a: &Pattern) -> u64 {
    let cols = rotate_pattern(a);
    return find_reflection(&cols, "vertical");
}

fn row_edit_distance(a: &Vec<char>, b: &Vec<char>) -> u64 {
    if a.len() != b.len() {
        panic!("mismatched lengths");
    }

    let mut edit_distance = 0;
    for i in 0..a.len() {
        if a[i] != b[i] {
            edit_distance += 1;
        }
    }
    return edit_distance;
}

fn find_smudged_reflection(a: &Pattern, direction: &str) -> u64 {
    use std::convert::TryFrom;

    for rowidx in 0..a.len() - 1 {
        let mut mirrorsize = 0;
        let mut mirror_edit_distance = 0;
        let mut x: u64;
        while mirrorsize <= rowidx && rowidx + mirrorsize + 1 < a.len() {
            x = row_edit_distance(&a[rowidx - mirrorsize], &a[rowidx + mirrorsize + 1]);
            if mirror_edit_distance + x > 1 {
                break;
            }
            mirror_edit_distance += x;
            mirrorsize += 1;
        }

        if mirrorsize == 0 || mirror_edit_distance != 1 {
            continue;
        }

        // It only counts if it goes up to an edge.
        if mirrorsize > rowidx && mirrorsize - rowidx == 1 || rowidx + mirrorsize + 1 == a.len() {
            //println!("found smudged reflection at {} of size {}", rowidx, mirrorsize);
            if direction == "horizontal" {
                return 100 * (u64::try_from(rowidx).unwrap() + 1);
            } else {
                return u64::try_from(rowidx).unwrap() + 1;
            }
        }
    }
    return 0;
}

fn find_smudged_horizontal_reflection(a: &Pattern) -> u64 {
    let val =  find_smudged_reflection(a, "horizontal");
    if val > 0 {
        println!("found smudged horizontal reflection, value: {}", val);
    }
    return val;
}

fn find_smudged_vertical_reflection(a: &Pattern) -> u64 {
    let cols = rotate_pattern(a);
    let val = find_smudged_reflection(&cols, "vertical");
    if val > 0 {
        println!("found smudged vertical reflection, value: {}", val);
    }
    return val;
}

fn do_pattern(a: &Pattern) -> u64 {
    //let mut val = find_horizontal_reflection(a);
    //val += find_vertical_reflection(a);
    let mut val = find_smudged_horizontal_reflection(a);
    val += find_smudged_vertical_reflection(a);
    return val;
}

fn main() {
    use std::fs::File;
    use std::io::{prelude::*, BufReader};

    let f = File::open("input.txt").unwrap();
    let reader = BufReader::new(f);

    let mut aidx = 0;
    let mut a: Pattern = vec![];
    let mut totalvalue: u64 = 0;

    for line in reader.lines() {
        let s = line.unwrap();
        if s == "" {
            println!("pattern {}", aidx);
            totalvalue += do_pattern(&a);
            a = vec![];
            aidx += 1;
            continue;
        }
        //println!("{}", s);
        let char_vec: Vec<char> = s.chars().collect();
        a.push(char_vec);
    }
    println!("pattern {}", aidx);
    totalvalue += do_pattern(&a);
    println!("total value: {}", totalvalue);
}
